hooks/pre-commit

```diff
@@ -1,9 +1,41 @@
-#! /bin/bash
+#!/bin/bash
 
 set -eu
 
+# make the GIT_DIR and GIT_INDEX_FILE absolute, before we change dir
+export GIT_DIR=$(realpath `git rev-parse --git-dir`)
+export GIT_INDEX_FILE=$(realpath `git rev-parse --git-path index`)
+
+wd=`pwd`
+
+# create a temp dir
+tmpdir=`mktemp -d`
+trap 'rm -rf "$tmpdir"' EXIT
+cd "$tmpdir"
+
+# get a copy of the index
+git checkout-index -a
+
+# run our checks
 golint src/...
 go fmt ./src/...
 go tool vet --shadow ./src
 gocyclo -over 12 src/
 gb test
+
+# if there are no changes from the index, we are done
+git diff --quiet && exit 0
+
+# we now need to apply any changes made to both the index and the working copy.
+# so, first get a patch
+git diff > "$GIT_DIR/pre-commit.patch"
+
+# add the changes to the index
+git add -u
+
+# attempt to apply the changes to the wc (but don't fail the commit for it if
+# there are conflicts).
```

Comment:
well, the commit itself will be ok; the only thing is that the changes made by `go fmt` won't be replicated to your working copy.

Suppose you have `git add`ed a file, and then made some more changes but not added them. When the pre-commit hook runs, we run `go fmt` against the staged version before it is committed; the problem is that the changes `go fmt` makes could conflict with the later changes you made in your working copy. I'm not sure what the correct behaviour would be in those circumstances, but I'd rather leave the working copy alone than mess it up. On the other hand, I see no reason that this should block the commit.